File: backend/app/schemas/community.py
"""
T3 커뮤니티 비교 API Pydantic schemas (P3-3).

6 endpoint:
- GET /api/v1/community/platforms/health
- GET /api/v1/community/platforms/product-matrix
- GET /api/v1/community/platforms/dispersion
- GET /api/v1/community/platforms/early-signal
- GET /api/v1/community/platforms/clusters
- GET /api/v1/community/platforms/anomalies

source MV:
- platform_health (1행/플랫폼)
- country_daily   (1행/day×country×product)
- voc_records     (라이브 — dispersion / early-signal 일부 계산용)
"""
from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


# ── 1) /health ─────────────────────────────────────────────────────
class PlatformHealth(BaseModel):
    platform_id: int
    code: str
    region: Optional[str] = None
    base_url: Optional[str] = None
    posts_24h: int
    posts_7d: int
    sent_avg_7d: Optional[float] = None
    avg_body_len_7d: Optional[int] = None
    last_collected: Optional[str] = None  # ISO timestamp
    status: str                            # 'active' | 'idle' | 'dead'


class PlatformHealthResponse(BaseModel):
    items: List[PlatformHealth]
    total: int
    active: int
    idle: int
    dead: int


# ── 2) /product-matrix ─────────────────────────────────────────────
class MatrixCell(BaseModel):
    platform: str
    product: str
    n: int
    sent_avg: float                        # -1 ~ 1
    neg_rate: float                        # 0 ~ 100


class ProductMatrixResponse(BaseModel):
    cells: List[MatrixCell]
    platforms: List[str]
    products: List[str]
    since: str


# ── 3) /dispersion ─────────────────────────────────────────────────
class BoxplotEntry(BaseModel):
    platform: str
    q1: float
    median: float
    q3: float
    iqr: float
    lo: float                              # whisker low (q1 - 1.5*iqr clipped to min)
    hi: float                              # whisker high
    n: int


class OutlierEntry(BaseModel):
    platform: str
    voc_id: int
    sentiment_score: float
    snippet: Optional[str] = None


class DispersionResponse(BaseModel):
    boxplot: List[BoxplotEntry]
    outliers: List[OutlierEntry]
    product_id: Optional[int] = None
    since: str


# ── 4) /early-signal ───────────────────────────────────────────────
class EarlySignalTimelinePoint(BaseModel):
    platform: str
    day: str                                # YYYY-MM-DD
    n: int
    sent_avg: float


class EarlySignalEvent(BaseModel):
    detected_at: str                        # YYYY-MM-DD
    leading_platform: Optional[str] = None
    lead_days: Optional[int] = None
    summary: str


class EarlySignalResponse(BaseModel):
    event: Optional[EarlySignalEvent] = None
    timeline: List[EarlySignalTimelinePoint]
    product_id: Optional[int] = None
    category: Optional[str] = None


# ── 5) /clusters ───────────────────────────────────────────────────
class ClusterPoint(BaseModel):
    platform: str
    cluster: int
    x: float                                # 2D 좌표 (PCA-lite: pos_rate vs neg_rate)
    y: float


class ClusterCentroid(BaseModel):
    cluster: int
    x: float
    y: float
    size: int                               # 멤버 개수
    pos_rate_avg: float
    neg_rate_avg: float
    sent_avg: float


class ClustersResponse(BaseModel):
    points: List[ClusterPoint]
    centroids: List[ClusterCentroid]
    k: int
    iterations: int


# ── 6) /anomalies ──────────────────────────────────────────────────
class AnomalyEntry(BaseModel):
    code: str                               # platform code
    reason: str                             # 'dead_7d' | 'idle_24h' | 'extreme_negative' | ...
    since: str                              # ISO timestamp or 'YYYY-MM-DD'
    detail: Dict[str, Any] = Field(default_factory=dict)


__all__ = [
    "PlatformHealth",
    "PlatformHealthResponse",
    "MatrixCell",
    "ProductMatrixResponse",
    "BoxplotEntry",
    "OutlierEntry",
    "DispersionResponse",
    "EarlySignalTimelinePoint",
    "EarlySignalEvent",
    "EarlySignalResponse",
    "ClusterPoint",
    "ClusterCentroid",
    "ClustersResponse",
    "AnomalyEntry",
]

File: frontend/src/types/community.ts
// T3 커뮤니티(플랫폼) 비교 페이지 타입 정의
// platform_health MV(72행) 및 country_daily MV(1016행) 기반의
// 비교/분산/조기신호/클러스터/이상치 분석을 위한 응답 스키마.

export type PlatformStatus = 'active' | 'idle' | 'dead';

// 1) 사이트 상태 — /platforms/health
export interface PlatformHealth {
  platform_id: number;
  code: string;
  region?: string | null;
  base_url?: string | null;
  posts_24h: number;
  posts_7d: number;
  sent_avg_7d: number | null;
  avg_body_len_7d: number | null;
  last_collected?: string | null;
  status: PlatformStatus;
}

export interface PlatformHealthResponse {
  platforms: PlatformHealth[];
  generated_at: string;
}

// 2) 제품 매트릭스 — /platforms/product-matrix
export interface ProductMatrixCell {
  platform_code: string;
  product_code: string;
  count: number;
  sent_avg: number;
}

export interface ProductMatrixResponse {
  cells: ProductMatrixCell[];
  platforms: string[];   // X축 (사이트 코드)
  products: string[];    // Y축 (제품 코드)
}

// 3) 분산 — /platforms/dispersion
// 각 플랫폼별 감성 분포 5수치 (boxplot 입력).
export interface DispersionEntry {
  platform_code: string;
  min: number;
  q1: number;
  median: number;
  q3: number;
  max: number;
  outliers?: number[];   // 시각화용 개별 이상치 점
  n: number;
}

export interface DispersionResponse {
  entries: DispersionEntry[];
}

// 4) Early Signal — /platforms/early-signal
// 특정 신호(=이슈/제품)가 각 플랫폼에 처음 등장한 시각의 lag (선두 대비 hours).
export interface EarlySignalRow {
  platform_code: string;
  signal: string;          // ex. 제품 코드 또는 이슈 라벨
  first_seen: string;      // ISO datetime
  lag_hours: number;       // 0 = 선두
  count_24h: number;       // 첫 등장 후 24h 누적
}

export interface EarlySignalResponse {
  signal: string;
  rows: EarlySignalRow[];
}

// 5) 클러스터 — /platforms/clusters
// 플랫폼을 2D 임베딩(예: PCA/UMAP 좌표) + 클러스터 ID 부여.
export interface ClusterPoint {
  platform_code: string;
  x: number;
  y: number;
  cluster: number;            // 0,1,2,...
  posts_7d: number;
  sent_avg_7d: number;
}

export interface ClusterResponse {
  points: ClusterPoint[];
  k: number;                  // 클러스터 개수
}

// 6) 이상치 — /platforms/anomalies
export type AnomalyKind = 'volume_spike' | 'volume_drop' | 'sent_swing' | 'silence';

export interface AnomalyItem {
  platform_code: string;
  kind: AnomalyKind;
  score: number;             // 0~1, 1에 가까울수록 강함
  detected_at: string;       // ISO datetime
  description: string;
}

export interface AnomalyResponse {
  items: AnomalyItem[];
}
